[DCA] Event collection from the API server

CharlyF opened the pull request:
### What does this PR do?

This PR introduces the event collection check.

### Motivation

The DCA needs to be able to collect events from the API server.
This introduces:
- [x] A method to connect to the API Server
- [x] A wrapper to collect the events
- [x] Leverage the new/modified outputs
- [x] Cache a map to properly aggregate related events
- [x] Format the events to have be properly processed by the intake
- [x] Map the events with the Human friendly InvolvedObject name
- [x] Flush the cache regularily
- [x] Add filtering
- [x] Add a function to collect and update a `LatestEventToken` (i.e. token of the most up to date version of the events collected by the DCA) from a ConfigMap `eventtokendca`. 
- [x] Add doc in the README.md and Dockerfiles/cluster-agent/README.md

### Testings

- [x] Image publicly available
- [x] Image tested
- [x] Integration tests
- [x] Unit tests

### Additional Notes
🐩

xvello committed:
add APIClient.LatestEvents() and integration test

Commit:
adding event aggregation formatting and submission

xvello commented on pkg/collector/corechecks/cluster/kubernetes_apiserver.go:
```diff
@@ -108,6 +109,32 @@ func (k *KubeASCheck) Run() error {
 		k.warn("could not collect API Server component status: ", err.Error())
 	}
+
+	newEvents, modifiedEvents, resVersion, err := asclient.LatestEvents(k.eventResVersion)
```
you should check `err != nil` and throw an integration warning before using the return values, then skip the whole section.

If the watcher fails, we return `since`, so your integration warning would never show up, see https://github.com/DataDog/datadog-agent/pull/1009/files#diff-8f1c08effc368ad3aa91fdd0007818dbR47

CharlyF replied:
just to clarify on this - You mean throw a warn with the err. As off now, because the err is not logged we don't log anything ? Not sure how to use `since`

xvello replied:
indeed, throw an integration warning, and skip the rest (update latestEvent, aggregate, send)

xvello commented on pkg/collector/corechecks/cluster/kubernetes_apiserver.go:
```diff
@@ -36,6 +36,7 @@ type KubeASCheck struct {
 	lastWarnings          []error
 	instance              *KubeASConfig
 	KubeAPIServerHostname string
+	eventResVersion       string
```
- Let's prefix the name by `latest` to make the intention clear.
- As the `APIClient.LatestEvents` doc does not mention the token is a `resVersion`, that implementation detail does not need to bleed into the check.

I'd go with `latestEvent` or `latestEventToken`, WDYT?

CharlyF replied:
yeah good point

xvello commented on pkg/collector/corechecks/cluster/kubernetes_apiserver.go:
```diff
@@ -108,6 +109,32 @@ func (k *KubeASCheck) Run() error {
 	if err != nil {
 		k.warn("could not collect API Server component status: ", err.Error())
 	}
+
```
Do you think the event collection logic should be behind a `collect_events` option?

CharlyF replied:
yeah eventually I want to add the option along with the filters on kind/type etc

xvello commented on pkg/collector/corechecks/cluster/kubernetes_apiserver.go:
```diff
@@ -108,6 +109,32 @@ func (k *KubeASCheck) Run() error {
+		k.eventResVersion = resVersion
+		evToSubmit, err := k.aggregateEvents(newEvents, false)
+		if err != nil {
+			k.warn("could not get the new events from the Kubenetes API Server", err.Error())
```
at this stage, the `err` is from the aggregate, not the apiserver

xvello commented on pkg/collector/corechecks/cluster/kubernetes_apiserver.go:
```diff
@@ -163,6 +190,27 @@ func (k *KubeASCheck) parseComponentStatus(sender aggregator.Sender, componentsS
 }
+
+func (k *KubeASCheck) aggregateEvents(events []*v1.Event, modified bool) (evs []metrics.Event, err error) {
+	eventsByNode := make(map[string]*kubernetesEventBundle)
```
s/eventsByNode/eventsByObject/ ?

xvello commented on pkg/collector/corechecks/cluster/kubernetes_apiserver.go:
```diff
@@ -163,6 +190,27 @@ func (k *KubeASCheck) parseComponentStatus(sender aggregator.Sender, componentsS
 	return nil
 }
+
+func (k *KubeASCheck) aggregateEvents(events []*v1.Event, modified bool) (evs []metrics.Event, err error) {
```
we prefer not to use the named return syntax for consistency

xvello reviewed (commented):
pretty solid! some lifecycle and readability improvements but the logic lgtm for now

Commit:
Formatting events, adding filters

CharlyF commented:
Improved format of events.
Relies on the object name as opposed to it's UID as some events would get duplicated in certain situations https://cl.ly/2h1s0v200h1F.
Also removed the component in the tags to avoid the aggregation conflict:
https://cl.ly/3v0l2o1R1q06 

Finally added a Filter type option:
```
2018-01-05 | DEBUG | (kubernetes_apiserver.go:212 in aggregateEvents) | filtered out the following events: 2 MissingClusterDNS
2018-01-05 | DEBUG | (kubernetes_apiserver.go:212 in aggregateEvents) | filtered out the following events: 2 MissingClusterDNS
```

Commit:
commit lock to pass tests

Commit:
fixing latest resversion lookup, adding filtered events, adding unit tests

Commit:
updating tests to be TZ agnostic

CharlyF commented:
Added Unit tests for new events, modified events, ignored events.
Added formatting https://cl.ly/1k38072r2f14 so events gather several kubernetes events from the same object. Using a similar format as Docker.
Fixed the resversion lookup so we make sure we keep the latest token as the basis for the next lookup.
Added the Ignore logic.

TODO:
- [ ] Add an init function acting as a placeholder for fetching the latest version token from a configmap.

Commit:
adding methods to fetch LatestEventToken from a Configmap at init. Also updates it

JulienBalestra commented on pkg/collector/corechecks/cluster/kubernetes_apiserver.go:
```diff
@@ -78,6 +83,51 @@ func (k *KubeASCheck) Run() error {
+			token = ""
+		}
+		k.configMapAvailable = found
+		k.latestEventToken, _ = strconv.Atoi(token)
```
We shouldn't ignore the error here.

JulienBalestra commented on pkg/collector/corechecks/cluster/kubernetes_apiserver.go:
```diff
@@ -78,6 +83,51 @@ func (k *KubeASCheck) Run() error {
+		k.Warn("could not collect events from the api server: ", err.Error())
+	}
+
+	lastVersion, _ := strconv.Atoi(versionToken)
```
ditto

JulienBalestra commented on pkg/collector/corechecks/cluster/kubernetes_apiserver.go:
```diff
@@ -78,6 +83,51 @@ func (k *KubeASCheck) Run() error {
+		err := k.aggregateEvents(sender, newEvents, false)
+
+		if err != nil {
+			k.Warnf("Could not submit new event", err.Error())
```
`Warnf` should have a formatter for the error like `%s`

JulienBalestra commented on pkg/collector/corechecks/cluster/kubernetes_apiserver.go:
```diff
@@ -78,6 +83,51 @@ func (k *KubeASCheck) Run() error {
+			err := k.aggregateEvents(sender, modifiedEvents, true)
+
+			if err != nil {
+				k.Warnf("Could not submit modified event ", err.Error())
```
ditto

JulienBalestra commented on pkg/collector/corechecks/cluster/kubernetes_apiserver_test.go:
```diff
@@ -16,6 +16,7 @@ import (
 	"github.com/DataDog/datadog-agent/pkg/aggregator/mocksender"
 	core "github.com/DataDog/datadog-agent/pkg/collector/corechecks"
 	"github.com/DataDog/datadog-agent/pkg/metrics"
+	"time"
```
You should run `goimports` here

JulienBalestra commented on pkg/collector/corechecks/cluster/kubernetes_apiserver_test.go:
```diff
@@ -120,3 +121,138 @@ func TestParseComponentStatus(t *testing.T) {
 
 	mock.AssertExpectations(t)
 }
+func createEvent(count int32, objname, objkind, objuid, component, reason string, message string, timestamp int64) *v1.Event {
```
you should add a newline between functions

JulienBalestra commented on pkg/collector/corechecks/cluster/kubernetes_apiserver_test.go:
```diff
@@ -120,3 +121,138 @@ func TestParseComponentStatus(t *testing.T) {
 	mock.AssertExpectations(t)
 }
+func createEvent(count int32, objname, objkind, objuid, component, reason string, message string, timestamp int64) *v1.Event {
+
```
You can remove this one

JulienBalestra commented on pkg/collector/corechecks/cluster/kubernetes_eventbundle.go:
```diff
@@ -0,0 +1,81 @@
+	"github.com/ericchiang/k8s/api/v1"
+	"math"
+	"strings"
+	"time"
```
`goimports` is needed here too

JulienBalestra commented on pkg/util/kubernetes/apiserver/apiserver.go:
```diff
@@ -87,20 +88,50 @@ func (c *APIClient) connect() error {
+func (c *APIClient) EventTokenFetcher() (string, bool, error) {
+	ctx, cancel := context.WithTimeout(context.Background(), c.timeout)
+	defer cancel()
+	tokenConfigMap, err := c.client.CoreV1().GetConfigMap(ctx, "eventtokendca", "default")
```
I think we could use a const to define **eventtokendca** and maybe the namespace.
It could be easier if we decide to make it configurable.

JulienBalestra commented on pkg/util/kubernetes/apiserver/apiserver.go:
```diff
@@ -87,20 +88,50 @@ func (c *APIClient) connect() error {
+	if err != nil {
+		return "", false, err
+	}
+	log.Infof("Fetched LatestEventToken from the eventtokendca ConfigMap")
```
Then here use a formatter to replace the **eventtokendca** by the new eventtokendca const

JulienBalestra commented on pkg/util/kubernetes/apiserver/apiserver.go:
```diff
@@ -87,20 +88,50 @@ func (c *APIClient) connect() error {
 	return c.client.CoreV1().ListComponentStatuses(ctx)
+}
 
+func (c *APIClient) EventTokenFetcher() (string, bool, error) {
```
Don't forget to comment public methods, I also encourage you to comment complex private methods too.

JulienBalestra commented on pkg/util/kubernetes/apiserver/apiserver.go:
```diff
@@ -87,20 +88,50 @@ func (c *APIClient) connect() error {
+	log.Debugf("LatestEventToken is %s", token)
+	return token, found, nil
+}
+func (c *APIClient) EventTokenSetter(token string) error {
```
you can add a newline and a comment for this public method.

JulienBalestra commented on pkg/util/kubernetes/apiserver/apiserver.go:
```diff
@@ -87,20 +88,50 @@ func (c *APIClient) connect() error {
+	if err != nil {
+		return err
+	}
+	log.Debugf("Updated LatestEventToken in the eventtokendca ConfigMap to %s", token)
```
This a relevant log line :+1:

JulienBalestra commented on pkg/util/kubernetes/apiserver/events.go:
```diff
@@ -0,0 +1,95 @@
+	"github.com/ericchiang/k8s/api/v1"
+)
+
+var eventReadTimeout = 100 * time.Millisecond
```
**eventReadTimeout** could be a const ?

CharlyF replied:
The benefit of using `var` is to force the type. It seems like something we want to enforce as early as possible.

JulienBalestra commented on pkg/util/kubernetes/apiserver/events.go:
```diff
@@ -0,0 +1,95 @@
+
+		if event != nil && event.Metadata != nil && event.Metadata.ResourceVersion != nil {
+			resVersionMetadata, err := strconv.Atoi(*event.Metadata.ResourceVersion)
+			resVersionCached, _ := strconv.Atoi(*latestResVersion)
```
`latestResVersion` come from string *since*, why ignoring the error ?

CharlyF replied:
So I figured this was something we initialised as a string and converted to a string before. I assumed an error would be redundant with:
https://github.com/DataDog/datadog-agent/blob/ac11dc23f8d0c70fe93beef305f60f866f19f3bb/pkg/collector/corechecks/cluster/kubernetes_apiserver.go#L97

xvello replied:
that logic might change sometime, might as well check indeed

JulienBalestra reviewed (changes requested):
This is a nice feature ! Big :+1: for the documentation
Few nits.

CharlyF commented on Dockerfiles/cluster-agent/README.md:
````diff
@@ -31,3 +31,100 @@ You'll need to download one of the `datadog-cluster-agent*_amd64.deb` package in
+```
+
+The ConfigMap to store the `LatestEventToken` has to be deployed in the `default` namespace and be named `eventtokendca`
+Simply running `kubectl create configmap eventtokendca --from-literal="eventToken"="12"`
````
You mean the yamls for the integrations ? kubernetes_apiserver.d/conf.yaml and co ?

JulienBalestra replied:
I was talking about the yaml for the configmap.

Commit:
review per @JulienBalestra's feedback

Commit:
updating functions in util to support generic interaction with CM

Commit:
catching error if cached resversion can't be converted

Commit:
update ConfigMapTokenSetter and ConfigMapTokenFetcher to use a timestamp associated with each token

Commit:
using '0' instead of '' for the resourceversion to use the apiserver local cache

CharlyF commented on test/integration/util/kube_apiserver/events_test.go:
```diff
@@ -48,7 +48,7 @@ func TestKubeEvents(t *testing.T) {
 
 	// Ignore potential startup events
-	_, _, initresversion, err := apiclient.LatestEvents("")
+	_, _, initresversion, err := apiclient.LatestEvents("0")
```
Forcing "0" instead of "", as we only want to use the APIServer's local cache.
if "" is passed, it panics:
```
panic: runtime error: invalid memory address or nil pointer dereference [recovered]
	panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x20 pc=0x10452d0]
```
The problem being with `resVersionCached, cachedResErr := strconv.Atoi(*latestResVersion)` [below](https://github.com/DataDog/datadog-agent/pull/1009/commits/3509c94605cbabb7c3bf67a77fc8505a2bc11a19#diff-8f1c08effc368ad3aa91fdd0007818dbR68). Although it returns 0 for the conversion (expected), it returns an error.

The solution I found would be to add `*latestResVersion=0` here:
https://github.com/DataDog/datadog-agent/pull/1009/commits/3509c94605cbabb7c3bf67a77fc8505a2bc11a19#diff-8f1c08effc368ad3aa91fdd0007818dbR41.

Thoughts @xvello, @JulienBalestra ?

JulienBalestra replied:
lgtm

xvello commented on pkg/collector/corechecks/cluster/kubernetes_apiserver.go:
```diff
@@ -122,6 +180,45 @@ func (k *KubeASCheck) parseComponentStatus(sender aggregator.Sender, componentsS
+
+ITER_EVENTS:
+	for _, event := range events {
+		for _, action := range k.instance.FilteredEventType {
```
Could you please add a comment "event type filtering" so the intention is explicit? In python we'd just write `if  action in FilteredEventType`, we loose readability in go :/

xvello commented on pkg/collector/corechecks/cluster/kubernetes_apiserver.go:
```diff
@@ -122,6 +180,45 @@ func (k *KubeASCheck) parseComponentStatus(sender aggregator.Sender, componentsS
+		}
+		bundle, found := eventsByObject[*event.InvolvedObject.Uid]
+		if found == false {
+			bundle = newKubernetesEventBundler(*event.InvolvedObject.Uid, *event.Source.Component)
```
are these pointers already tested as non `nil` in the apiclient, or should we check here before dereferencing?

CharlyF replied:
This one is not nil indeed http://kubernetes.io/docs/user-guide/identifiers#uids

xvello commented on pkg/collector/corechecks/cluster/kubernetes_eventbundle.go:
```diff
@@ -0,0 +1,82 @@
+func newKubernetesEventBundler(objUid string, compName string) *kubernetesEventBundle {
+	return &kubernetesEventBundle{
+		objUid:        objUid,
+		events:        []*v1.Event{},
```
iirc arrays already have a sane default value, you can skip this

xvello commented on pkg/collector/corechecks/cluster/kubernetes_eventbundle.go:
```diff
@@ -0,0 +1,82 @@
+		return fmt.Errorf("mismatching Object UIDs: %s != %s", event.InvolvedObject.Uid, k.objUid)
+	}
+	k.events = append(k.events, event)
+	k.countByAction[fmt.Sprintf("**%s**: %s\n", *event.Reason, *event.Message)] += int(*event.Count)
```
Same here, `nil` pointer dereference waiting to happen :/

xvello commented on pkg/collector/corechecks/cluster/kubernetes_eventbundle.go:
```diff
@@ -0,0 +1,82 @@
+	}
+	k.events = append(k.events, event)
+	k.countByAction[fmt.Sprintf("**%s**: %s\n", *event.Reason, *event.Message)] += int(*event.Count)
+	k.timeStamp = int64(math.Max(float64(k.timeStamp), float64(*event.Metadata.CreationTimestamp.Seconds)))
```
let's keep `k.timeStamp` and `k.lastTimestamp` as float64 to avoid re-casting everytime

xvello commented on pkg/collector/corechecks/cluster/kubernetes_eventbundle.go:
```diff
@@ -0,0 +1,82 @@
+	k.countByAction[fmt.Sprintf("**%s**: %s\n", *event.Reason, *event.Message)] += int(*event.Count)
+	k.timeStamp = int64(math.Max(float64(k.timeStamp), float64(*event.Metadata.CreationTimestamp.Seconds)))
+	k.lastTimestamp = int64(math.Max(float64(k.timeStamp), float64(*event.LastTimestamp.Seconds)))
+	k.readableKey = fmt.Sprintf("%s %s", *event.InvolvedObject.Name, *event.InvolvedObject.Kind)
```
Do we need to compute it for every event we add to the bundle?

You could also store the name and kind, and format it in the `formatEvents` method, instead of `Sprintf`ing here

CharlyF replied:
Yeah, I prefer this approach - For every New or Modified event, we create these keys instead of surcharging the bundle with `*event.InvolvedObject.Name, *event.InvolvedObject.Kind` 
Also, we build the countByAction map at this moment (it's a per event computation) and we need `*event.Reason, *event.Message`. I figured it would be a good balance to have it this way.

Commit:
moviong strconv of the local latestresver out of the watcher loop

xvello commented on pkg/collector/corechecks/cluster/kubernetes_eventbundle.go:
```diff
@@ -0,0 +1,82 @@
+		AggregationKey: fmt.Sprintf("kubernetes_apiserver:%s", k.objUid),
+	}
+	if len(k.events) == 0 {
+		return output, errors.New("no event to export")
```
You should check `k.events` lenght first, and just return `metrics.Event{}, error`

xvello commented on pkg/collector/corechecks/cluster/kubernetes_eventbundle.go:
```diff
@@ -0,0 +1,82 @@
+	}
+
+	output.Title = fmt.Sprintf("Events from the %s",
+		k.readableKey)
```
no need for newline here

xvello commented on pkg/collector/corechecks/cluster/kubernetes_apiserver.go:
```diff
@@ -71,12 +78,63 @@ func (k *KubeASCheck) Run() error {
+
+	var token string
+
+	if k.latestEventToken == 0 {
```
You can move the ConfigMapTokenFetcher call inside the `Config()` method, and don't need to check `k.latestEventToken == 0` at every run

xvello commented on pkg/util/kubernetes/apiserver/apiserver.go:
```diff
@@ -87,20 +97,79 @@ func (c *APIClient) connect() error {
+}
+
+// ConfigMapTokenFetcher returns the value of the `tokenValue` from the `tokenKey` in the ConfigMap `configmaptokendca` if its timestamp is less than tokenTimeout old.
+func (c *APIClient) ConfigMapTokenFetcher(tokenKey string) (string, bool, error) {
```
Let's make a backlog card about it: I'd prefer to get a `ConfigMap` object from the apiclient and call it's `Get`/`Set` methods, like the `DockerListener` creates `Services` that provide methods.

Later we'll be able to derive an interface other orchestrators can provide for value persistence.

- For now, just remove references to tokens, we store/retrieve a value in a configmap name

- The name is not great, if the suffix is a substantive (`Fetcher`) we usually assume it's an object/class, method names usually are verbs

xvello commented on pkg/util/kubernetes/apiserver/apiserver.go:
```diff
@@ -9,20 +9,30 @@ package apiserver
+	configMapDCAToken = "configmaptokendca"
+	defaultNamespace  = "default"
+	tokenTime         = "tokenTimestamp"
+	tokenTimeout      = 60
```
This is specific to event collection, other application will require different freshness values (or event won't care). This should be passed as an argument

xvello commented on pkg/util/kubernetes/apiserver/apiserver.go:
```diff
@@ -87,20 +97,79 @@ func (c *APIClient) connect() error {
+	defer cancel()
+	tokenConfigMap, err := c.client.CoreV1().GetConfigMap(ctx, configMapDCAToken, defaultNamespace)
+	if err != nil {
+		return "", false, err
```
`false` is pretty non-descript, I would've gone with a standard error clients can check and ignore, like https://github.com/DataDog/datadog-agent/blob/84538338fd2e8210e20b46a2d8747afea571b4c7/pkg/tagger/collectors/types.go#L25

xvello replied:
you could also create an error type for when the configmap is here but too old

xvello commented on pkg/util/kubernetes/apiserver/apiserver.go:
```diff
@@ -87,20 +97,79 @@ func (c *APIClient) connect() error {
+	if err != nil {
+		return "", false, err
+	}
+	log.Infof("Found the ConfigMap %s", configMapDCAToken)
```
spammy

CharlyF replied:
This will only show up once, when the agent starts. I think info is legit

xvello commented on pkg/util/kubernetes/apiserver/apiserver.go:
```diff
@@ -87,20 +97,79 @@ func (c *APIClient) connect() error {
+	if !found {
+		return "", found, log.Errorf("%s was not found in the ConfigMap %s", tokenKey, configMapDCAToken)
+	}
+	log.Debugf("%s is %s", tokenKey, token)
```
spammy

CharlyF replied:
Moved to trace

xvello commented on pkg/util/kubernetes/apiserver/apiserver.go:
```diff
@@ -87,20 +97,79 @@ func (c *APIClient) connect() error {
+	}
+	log.Debugf("%s is %s", tokenKey, token)
+
+	tokenTimeStr, set := tokenConfigMap.Data[fmt.Sprintf("%s.%s", tokenKey, tokenTime)] // This is so we can have one timestamp per tokenKey
```
why not simply have `tokenConfigMap.Data["value"]` and `tokenConfigMap.Data["timestamp"]` instead of `Sprintf`ing this key?

CharlyF replied:
As `ConfigMapTokenFetcher` and `ConfigMapTokenSetter` are meant to be as agnostic as possible, the idea is to give a key and the method does the job to store/query the right key name and the timestamp associated.
The idea for instance would be to store the event key token along with the up to date timestamp and as a second entry in the configmap, the leader elected uid and the timestamp of the last update.

xvello commented on pkg/util/kubernetes/apiserver/apiserver.go:
```diff
@@ -87,20 +97,79 @@ func (c *APIClient) connect() error {
+	tokenTimeStr, set := tokenConfigMap.Data[fmt.Sprintf("%s.%s", tokenKey, tokenTime)] // This is so we can have one timestamp per tokenKey
+
+	if set {
+		tokenTime, err := strconv.Atoi(tokenTimeStr)
```
you should use the `time` package's methods instead of converting on your own. See https://play.golang.org/p/ubuGVolITGb

xvello commented on pkg/util/kubernetes/apiserver/apiserver.go:
```diff
@@ -87,20 +97,79 @@ func (c *APIClient) connect() error {
+		tokenAge := int(time.Now().Unix()) - tokenTime
+
+		if tokenAge > tokenTimeout {
+			log.Infof("The event token is outdated, refreshing the events")
```
`refreshing the events` indicates your logic is too specific to event collection :(

CharlyF replied:
Correct 👍

Commit:
improving timestamp logic in the CM

CharlyF commented on pkg/collector/corechecks/cluster/kubernetes_apiserver.go:
```diff
@@ -122,6 +180,45 @@ func (k *KubeASCheck) parseComponentStatus(sender aggregator.Sender, componentsS
 	return nil
 }
+
+func (k *KubeASCheck) aggregateEvents(sender aggregator.Sender, events []*v1.Event, modified bool) error {
```
I propose process events

Commit:
using string logic instead of int for resversion agnosticity

Commit:
improving logic, using better variable types

Commit:
last improvements: token, errors, better returns

Commit:
return token when outdated

xvello commented on pkg/collector/corechecks/cluster/kubernetes_eventbundle.go:
```diff
@@ -0,0 +1,84 @@
+	if event.Reason == nil || event.Message == nil || event.InvolvedObject.Name == nil || event.InvolvedObject.Kind == nil {
+		return errors.New("could not retrieve some attributes of the event")
+	}
+	k.countByAction[fmt.Sprintf("**%s**: %s\n", *event.Reason, *event.Message)] += int(*event.Count)
```
IIRC from my testing, `event.Count` holds the total number of repetitions from the creation of the event.

That means if we got repetitions 4 to 8 in a bucket, we'll have 5 events with counts [4, 5, 6, 7, 8]. I fear `k.countByAction` will hold 30 instead of 5. We should test that

CharlyF replied:
If we have a bucket of N events with attributes generating the same aggregation key, the counts will indeed be added and it'll result in Datadog_Event: (Sum(event.count[i], i from 4 to 8) event.Reason event.Message) and this is the desired behaviour in my opinion.

This is what it end up looking like:
https://cl.ly/2D3Q342o0N2v

As we aggregate per ObjUid we care about the underlying counts being aggregated not the number of occurrence of an event during 15 seconds.

Commit:
document eventBundle struct. Avoid nil pointer evaluation

JulienBalestra commented on pkg/collector/corechecks/cluster/kubernetes_apiserver.go:
```diff
@@ -71,12 +80,60 @@ func (k *KubeASCheck) Run() error {
+				k.configMapAvailable = found
+				k.latestEventToken = token
+			}
+
```
Can we remove this line ?

JulienBalestra commented on pkg/collector/corechecks/cluster/kubernetes_apiserver_test.go:
```diff
@@ -120,3 +122,138 @@ func TestParseComponentStatus(t *testing.T) {
+	kubeASCheck.processEvents(mock, filteredKubeEventsBundle, false)
+	mock.AssertNotCalled(t, "Event")
+	mock.AssertExpectations(t)
+
```
Can we remove this empty line ?

JulienBalestra commented on pkg/util/kubernetes/apiserver/apiserver.go:
```diff
@@ -9,20 +9,30 @@ package apiserver
 
 import (
 	"context"
```
Are theses imports formatted ?

CharlyF replied:
missing a line indeed

JulienBalestra commented on pkg/util/kubernetes/apiserver/apiserver.go:
```diff
@@ -87,20 +97,82 @@ func (c *APIClient) connect() error {
+}
+
+// ConfigMapTokenFetcher returns the value of the `tokenValue` from the `tokenKey` in the ConfigMap `configmaptokendca` if its timestamp is less than tokenTimeout old.
+func (c *APIClient) ConfigMapTokenFetcher(token string, tokenTimeout int64) (string, bool, error) {
```
I'll be more confortable with a public method name like `GetTokenFromConfigmap`
Is it possible to explain how the custom errors are working ?

CharlyF replied:
Also updated `ConfigMapTokenSetter` to `UpdateTokenInConfigmap`

xvello reviewed (dismissed):
LGTM for the logic, I'll leave the final approval to @JulienBalestra once his comments are addressed

Commit:
improving format and method name per @JulienBalestra's review

JulienBalestra commented on pkg/collector/corechecks/cluster/kubernetes_apiserver.go:
```diff
@@ -71,12 +80,59 @@ func (k *KubeASCheck) Run() error {
+		k.Warnf("Could not collect API Server component status: %s", err.Error())
+	}
+
+	if k.instance.CollectEvent {
```
We can failfast here if it's false

JulienBalestra commented on pkg/collector/corechecks/cluster/kubernetes_apiserver.go:
```diff
@@ -71,12 +80,59 @@ func (k *KubeASCheck) Run() error {
+			k.Warnf("Could not collect events from the api server: %s", err.Error())
+		}
+		// We check that the resversion gotten from the API Server is more recent than the one cached in the util.
+		if len(newEvents)+len(modifiedEvents) > 0 {
```
we can return fast here too

JulienBalestra commented on pkg/collector/corechecks/cluster/kubernetes_apiserver.go:
```diff
@@ -71,12 +80,59 @@ func (k *KubeASCheck) Run() error {
+				}
+			}
+			err := k.processEvents(sender, newEvents, false)
+
```
I may prefer stick the error management with to the function/method call

JulienBalestra commented on pkg/collector/corechecks/cluster/kubernetes_apiserver.go:
```diff
@@ -71,12 +80,59 @@ func (k *KubeASCheck) Run() error {
+			}
+
+			// We send the events in 2 steps to make sure the new events are initializing the aggregation keys and as modified events have a different payload.
+			if len(modifiedEvents) > 0 {
```
Continue to return fast here

JulienBalestra commented on pkg/collector/corechecks/cluster/kubernetes_apiserver.go:
```diff
@@ -71,12 +80,59 @@ func (k *KubeASCheck) Run() error {
+			// We send the events in 2 steps to make sure the new events are initializing the aggregation keys and as modified events have a different payload.
+			if len(modifiedEvents) > 0 {
+				err := k.processEvents(sender, modifiedEvents, true)
+
```
ditto for the stick

JulienBalestra commented on pkg/collector/corechecks/cluster/kubernetes_apiserver.go:
```diff
@@ -122,6 +178,47 @@ func (k *KubeASCheck) parseComponentStatus(sender aggregator.Sender, componentsS
+		}
+
+		if len(filteredByType) > 0 {
+			log.Debugf("filtered out the following events: %s", formatStringIntMap(filteredByType))
```
Capitalise the log messages

JulienBalestra commented on pkg/collector/corechecks/cluster/kubernetes_apiserver.go:
```diff
@@ -122,6 +178,47 @@ func (k *KubeASCheck) parseComponentStatus(sender aggregator.Sender, componentsS
 	return nil
 }
+
+// processEvents aggregates and submits the events collected by the API Server.
```
```go
// processEvents:
- iterates on the Kubernetes Events
- extracts some attributes and builds a structure ready to be submitted as a Datadog event (bundle)
- formats the bundle and submit the Datadog event```

JulienBalestra commented on pkg/collector/corechecks/cluster/kubernetes_eventbundle.go:
```diff
@@ -0,0 +1,87 @@
+	"github.com/ericchiang/k8s/api/v1"
+)
+
+type kubernetesEventBundle struct {
```
nice 👍

JulienBalestra commented on pkg/collector/corechecks/cluster/kubernetes_eventbundle.go:
```diff
@@ -0,0 +1,87 @@
+	k.timeStamp = math.Max(k.timeStamp, float64(*event.Metadata.CreationTimestamp.Seconds))
+	k.lastTimestamp = math.Max(k.timeStamp, float64(*event.LastTimestamp.Seconds))
+
+	if event.Reason == nil || event.Message == nil || event.InvolvedObject.Name == nil || event.InvolvedObject.Kind == nil {
```
Is it possible to move this on top ? Do we need to update k ?

CharlyF replied:
let's ignore the events that are not properly formatted and also ignore their contribution.
i.e. : if a new event is misformated, we don't want to have part of it's message in the countByAction for instance.

JulienBalestra commented on pkg/collector/corechecks/cluster/kubernetes_eventbundle.go:
```diff
@@ -0,0 +1,87 @@
+		AggregationKey: fmt.Sprintf("kubernetes_apiserver:%s", k.objUid),
+	}
+
+	output.Title = fmt.Sprintf("Events from the %s", k.readableKey)
```
Can you set the title inside the init ?

JulienBalestra commented on pkg/util/kubernetes/apiserver/apiserver.go:
```diff
@@ -87,20 +98,82 @@ func (c *APIClient) connect() error {
+	defer cancel()
+	tokenConfigMap, err := c.client.CoreV1().GetConfigMap(ctx, configMapDCAToken, defaultNamespace)
+	if err != nil {
+		log.Debugf("could not find the ConfigMap %s: %s", configMapDCAToken, err.Error())
```
s/could/Could

JulienBalestra commented on pkg/util/kubernetes/apiserver/apiserver.go:
```diff
@@ -87,20 +98,82 @@ func (c *APIClient) connect() error {
+
+	tokenTimeStr, set := tokenConfigMap.Data[fmt.Sprintf("%s.%s", token, tokenTime)] // This is so we can have one timestamp per token
 
+	if set {
```
Can we return fast here if !set ?

JulienBalestra commented on pkg/util/kubernetes/apiserver/events.go:
```diff
@@ -0,0 +1,104 @@
+		}
+		timeout.Reset(eventReadTimeout)
+
+		if event != nil && event.Metadata != nil && event.Metadata.ResourceVersion != nil {
```
We can log.Debug and continue if anything is nil here

JulienBalestra commented on pkg/util/kubernetes/apiserver/events.go:
```diff
@@ -0,0 +1,104 @@
+			continue
+		}
+
+		if meta != nil && meta.Type != nil {
```
ditto

JulienBalestra reviewed (changes requested):
some nits on code style, comments.

mfpierre commented on Dockerfiles/cluster-agent/README.md:
```diff
@@ -31,3 +31,100 @@ You'll need to download one of the `datadog-cluster-agent*_amd64.deb` package in
+apiVersion: extensions/v1beta1
+kind: Deployment
+metadata:
+  name: dca
```
maybe we should have a more explicit name or at least `dd-dca` or `datadog-dca`?

CharlyF replied:
that's reasonable 👍

mfpierre commented on pkg/util/kubernetes/apiserver/apiserver.go:
```diff
@@ -9,20 +9,31 @@ package apiserver
 var globalApiClient *APIClient
 
+const (
+	configMapDCAToken = "configmaptokendca"
```
it bugs me that words are not in the same order, maybe we should rename `configmaptokendca` to `configmapdcatoken`

CharlyF replied:
🐛  accepted

Commit:
serious improvments

Commit:
imrpove the documentation

Commit:
fix if conditions

CharlyF commented:
💯
